### ISSUE TITLE
SSH over OpenVPN connection fails when Docker Swarm is running

### ISSUE THREAD

**Issue by joelchen** [moby/moby#31972]
**Description**

Connection timed out when I ssh into my AWS instance through OpenVPN connection after `docker swarm init`.

**Steps to reproduce the issue:**
1. Launch and install Docker onto 2 AWS instances on the same network (10.0.0.1 and 10.0.0.2).
2. Run OpenVPN (https://github.com/kylemanna/docker-openvpn) on instance A (10.0.0.1).
3. Run `docker swarm init` on instance B (10.0.0.2).
4. Connect to AWS network through OpenVPN of instance A (10.0.0.1).
5. Run `ssh -i ~/.ssh/key.pem ubuntu@10.0.0.2` from my local machine.

**Describe the results you received:**
`ssh: connect to host 10.0.0.2 port 22: Operation timed out`

**Describe the results you expected:**
I expected ssh through OpenVPN to login successfully into instance B (10.0.0.2).
```
Last login: Tue Mar 21 15:04:40 2017 from ip-10-0-0-1.ap-southeast-1.compute.internal
ubuntu@ip-10-0-0-2:~$
```

**Additional information you deem important (e.g. issue happens only occasionally):**
Issue occurs even after ports described in https://docs.docker.com/engine/swarm/swarm-tutorial/#open-protocols-and-ports-between-the-hosts are opened. Able to ssh through OpenVPN into instance B when Docker Swarm is not running on instance B. Able to ssh through AWS Internet Gateway (if configured) into instance B when Docker Swarm is running on instance B. Without OpenVPN, still unable to ssh (connection timed out) into instance B through instance A (ssh into instance A through Internet Gateway first) when Docker Swarm is running on instance B.

OpenVPN configuration (openvpn.conf):
```
server 192.168.255.0 255.255.255.0
verb 3
key /etc/openvpn/pki/private/<masked public IP>.key
ca /etc/openvpn/pki/ca.crt
cert /etc/openvpn/pki/issued/<masked public IP>.crt
dh /etc/openvpn/pki/dh.pem
tls-auth /etc/openvpn/pki/ta.key
key-direction 0
keepalive 10 60
persist-key
persist-tun

proto udp
# Rely on Docker to do port mapping, internally always 1194
port 1194
dev tun0
status /tmp/openvpn-status.log

user nobody
group nogroup

### Route Configurations Below
route 192.168.254.0 255.255.255.0

### Push Configurations Below
push block-outside-dns
push dhcp-option DNS 8.8.8.8
push dhcp-option DNS 8.8.4.4

### Extra Configurations Below
```

**Output of `docker version`:**

```
ubuntu@ip-10-0-0-2:~$ docker version
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:57:47 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:57:47 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
ubuntu@ip-10-0-0-2:~$ docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 17.03.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 0
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: mcsklxrf8a1au6jpv1pk3z36e
 Is Manager: true
 ClusterID: iqwn12hqnuey6qs3uh2nkooyf
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.0.0.2
 Manager Addresses:
  10.0.0.2:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Security Options:
 apparmor
Kernel Version: 3.13.0-112-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.676 GiB
Name: ip-10-254-161-137
ID: J7V7:TEXJ:USIZ:GYII:VFTL:5LKS:7YUM:LEZE:JLJK:FHNZ:QKXJ:2W6T
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
AWS environment.

**Comment by aluzzardi** [comment]
Not sure how this is related to Swarm.

Might we be setting some overlay iptables rule, @mavenugo @sanimej @aboch?

**Comment by joelchen** [comment]
Without OpenVPN, this issue (connection timed out) is still occurring when I ssh into instance B through instance A (ssh into instance A through Internet Gateway first) when Docker Swarm is running on instance B.

**Comment by fcrisciani** [comment]
Hi @joelchen is your issue consistently reproducible?
I tried to reproduce with a couple of vms locally and the ssh for me works between the 2 vms. Is it possible that there is some AWS config that is not allowing the traffic?

Also if you have the issue reproduced, will you be able to run a tcpdump on the sender and receiver side so to confirm that the TCP handshake happens of in case which side is not responding?

**Comment by joelchen** [comment]
@fcrisciani Are your VMs running Ubuntu 14.04? There are too many outputs from tcpdump, how do I spot the TCP handshake?

Here is out put of iptables before `docker swarm init`:
```
ubuntu@ip-10-0-0-2:~$ sudo iptables -nvL
Chain INPUT (policy ACCEPT 128 packets, 8239 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
 332K   60M DOCKER-ISOLATION  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 DOCKER-OVERLAY  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0

Chain OUTPUT (policy ACCEPT 92 packets, 9958 bytes)
 pkts bytes target     prot opt in     out     source               destination
89922 8019K DOCKER-OVERLAY  all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain DOCKER (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain DOCKER-ISOLATION (1 references)
 pkts bytes target     prot opt in     out     source               destination
3380K  613M RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain DOCKER-OVERLAY (2 references)
 pkts bytes target     prot opt in     out     source               destination
 137K   13M RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0
```
Here is out put of iptables after `docker swarm init`:
```
ubuntu@ip-10-0-0-2:~$ sudo iptables -nvL
Chain INPUT (policy ACCEPT 23 packets, 2011 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 DOCKER-ISOLATION  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 DOCKER     all  --  *      docker_gwbridge  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      docker_gwbridge  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  docker_gwbridge !docker_gwbridge  0.0.0.0/0            0.0.0.0/0
    0     0 DOCKER-OVERLAY  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
    0     0 DROP       all  --  docker_gwbridge docker_gwbridge  0.0.0.0/0            0.0.0.0/0

Chain OUTPUT (policy ACCEPT 18 packets, 1914 bytes)
 pkts bytes target     prot opt in     out     source               destination
   18  1914 DOCKER-OVERLAY  all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain 9zxr6jkl5b7g (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     all  --  ov-001000-9zxr6 *       0.0.0.0/0            0.0.0.0/0
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain DOCKER (2 references)
 pkts bytes target     prot opt in     out     source               destination

Chain DOCKER-ISOLATION (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 DROP       all  --  docker0 docker_gwbridge  0.0.0.0/0            0.0.0.0/0
    0     0 DROP       all  --  docker_gwbridge docker0  0.0.0.0/0            0.0.0.0/0
3380K  613M RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain DOCKER-OVERLAY (2 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 9zxr6jkl5b7g  all  --  *      ov-001000-9zxr6  0.0.0.0/0            0.0.0.0/0
 137K   13M RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0
```

**Comment by joelchen** [comment]
Upgrading to latest Ubuntu 16.04, ssh through VPN or through another instance works and I am unable to reproduce this issue. Can Docker team fix this issue with Ubuntu 14.04?

Here is output of iptables on Ubuntu 16.04 before `docker swarm init`:
```
ubuntu@ip-10-0-0-2:~$ sudo iptables -nvL
Chain INPUT (policy ACCEPT 34 packets, 2319 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 DOCKER-ISOLATION  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0

Chain OUTPUT (policy ACCEPT 27 packets, 2910 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain DOCKER (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain DOCKER-ISOLATION (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0
```
Here is output of iptables on Ubuntu 16.04 after `docker swarm init`:
```
ubuntu@ip-10-0-0-2:~$ sudo iptables -nvL
Chain INPUT (policy ACCEPT 8 packets, 703 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 DOCKER-ISOLATION  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 DOCKER     all  --  *      docker_gwbridge  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      docker_gwbridge  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  docker_gwbridge !docker_gwbridge  0.0.0.0/0            0.0.0.0/0
    0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
    0     0 DROP       all  --  docker_gwbridge docker_gwbridge  0.0.0.0/0            0.0.0.0/0

Chain OUTPUT (policy ACCEPT 5 packets, 510 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain DOCKER (2 references)
 pkts bytes target     prot opt in     out     source               destination

Chain DOCKER-ISOLATION (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 DROP       all  --  docker0 docker_gwbridge  0.0.0.0/0            0.0.0.0/0
    0     0 DROP       all  --  docker_gwbridge docker0  0.0.0.0/0            0.0.0.0/0
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0
```

**Comment by fcrisciani** [comment]
@joelchen I tried to spawn an Ubuntu 14.04 on my laptop and after the docker swarm init looks like I'm still able to ssh to it.

Also the iptables output that you posted are looking good, don't see any DROP counter positive, meaning that right now there is no iptables rule preventing the ssh traffic.

Following I'm adding the tcpdump command to use to see if the packets are actually being sent and received on the interfaces.

for the tcpdump command you can use the following, launch both sides before attempting a new ssh connection. (consider that if you have one already active between the 2 vms the output will be pretty busy, so check that before starting the command is launch but no output is generated):

**Source VM (machine that start the ssh command):**
**sudo tcpdump -eni <interface name, usually eth0> port 22 and host <ip address of the destination vm>**

The output that we want to see on this VM if the TCP connection is established is the following (these are the very first 3 lines printed by the command):

`14:56:29.096662 08:00:27:f9:f6:b5 > 08:00:27:f9:f6:2f, ethertype IPv4 (0x0800), length 78: 192.168.201.169.50826 > 192.168.201.72.22: Flags [S], seq 1252758297, win 65535, options [mss 1460,nop,wscale 5,nop,nop,TS val 402839027 ecr 0,sackOK,eol], length 0`

`14:56:29.096967 08:00:27:f9:f6:2f > 08:00:27:f9:f6:b5, ethertype IPv4 (0x0800), length 74: 192.168.201.72.22 > 192.168.201.169.50826: Flags [S.], seq 10832244, ack 1252758298, win 28960, options [mss 1460,sackOK,TS val 247864 ecr 402839027,nop,wscale 7], length 0`

`14:56:29.097036 08:00:27:f9:f6:b5 > 08:00:27:f9:f6:2f, ethertype IPv4 (0x0800), length 66: 192.168.201.169.50826 > 192.168.201.72.22: Flags [.], ack 1, win 4117, options [nop,nop,TS val 402839027 ecr 247864], length 0`
`[...] lot of other stuff if the connection goes through, we can ignore them for the moment`


**Destination VM:**
**sudo tcpdump -eni <interface name, usually eth0> port 22 and host <ip address of the source vm>**

Also here the expected output is:
`15:05:01.476726 08:00:27:f9:f6:b5 > 08:00:27:f9:f6:2f, ethertype IPv4 (0x0800), length 78: 192.168.201.169.50909 > 192.168.201.72.22: Flags [S], seq 384006490, win 65535, options [mss 1460,nop,wscale 5,nop,nop,TS val 403348531 ecr 0,sackOK,eol], length 0`

`15:05:01.476807 08:00:27:f9:f6:2f > 08:00:27:f9:f6:b5, ethertype IPv4 (0x0800), length 74: 192.168.201.72.22 > 192.168.201.169.50909: Flags [S.], seq 3375471505, ack 384006491, win 28960, options [mss 1460,sackOK,TS val 375966 ecr 403348531,nop,wscale 7], length 0`

`15:05:01.476948 08:00:27:f9:f6:b5 > 08:00:27:f9:f6:2f, ethertype IPv4 (0x0800), length 66: 192.168.201.169.50909 > 192.168.201.72.22: Flags [.], ack 1, win 4117, options [nop,nop,TS val 403348531 ecr 375966], length 0`
`[...] other lines that we can ignore`

**Comment by joelchen** [comment]
From instance B (Ubuntu 14.04) running Docker Swarm:
```
06:03:30.090998 IP 10.0.0.1.41466 > 10.0.0.2.22: Flags [S], seq 784855174, win 26883, options [mss 8961,sackOK,TS val 74728763 ecr 0,nop,wscale 7], length 0
06:03:32.095177 IP 10.0.0.1.41466 > 10.0.0.2.22: Flags [S], seq 784855174, win 26883, options [mss 8961,sackOK,TS val 74729264 ecr 0,nop,wscale 7], length 0
06:03:35.983839 ARP, Request who-has 10.0.0.2 tell 10.0.0.2, length 42
06:03:35.983863 ARP, Reply 10.0.0.2 is-at 02:3d:5b:51:cb:ad, length 28
```
From instance A (Ubuntu 14.04/16.04) trying to ssh into instance B (Ubuntu 14.04):
```
05:56:07.117611 IP 10.0.0.1.41464 > 10.0.0.2.22: Flags [S], seq 2640496760, win 26883, options [mss 8961,sackOK,TS val 74624038 ecr 0,nop,wscale 7], length 0
05:56:08.117461 IP 10.0.0.1.41464 > 10.254.161.105.22: Flags [S], seq 2640496760, win 26883, options [mss 8961,sackOK,TS val 74624288 ecr 0,nop,wscale 7], length 0
05:56:08.248992 ARP, Request who-has 10.0.0.1 tell 10.0.0.1, length 42
05:56:08.249011 ARP, Reply 10.0.0.1 is-at 02:85:9d:73:f7:f9, length 28
```

**Comment by fcrisciani** [comment]
Hi @joelchen, in kernel 3.13.X, that you are currently using, there are some limitations where the overlay network vxlan interfaces are created in the host namespace. 
In this case as you do the docker swarm init the network 10.0.0.0/8 is starting being used by docker.
Try changing the host network that you are using in AWS so that your host interface of the vms that you are using are not anymore on the 10.0.0.0/8. That should solve the problem for you in case of vms with old kernel versions.
Let me know if it works.

**Comment by joelchen** [comment]
@fcrisciani It is actually an instance of VPC with subnet 10.254.XXX.XXX. I replaced it with 10.0.0.X for illustration. Unfortunately, it falls within 10.0.0.0/8 CIDR, and I must use this network for my project.

**Comment by mavenugo** [comment]
@joelchen as @fcrisciani explained, the issue is with the 3.13 kernel and the overlapping subnet issue with an overlay network. In swarm-mode, we create a default overlay network named `ingress` with a subnet of `10.255.x.x/16` and that might interfere with your subnet.

FYI, we added an ability to remove the `ingress` network via https://github.com/docker/docker/pull/31714 and one can create the ingress network in any other non-conflicting subnet while running in the older kernels.

This feature will be available for 17.05 release and if you are interested in trying it out, pls try a master build.

**Comment by joelchen** [comment]
@mavenugo Nice to know that there is a way to overcome this issue. Can the removal be performed before `docker swarm init`? If it can only be done after `docker swarm init`, what happens will be the same, I will be kicked out of the SSH session, and I have to find another way (attaching Internet Gateway) to access the instance to do `docker network rm ingress` without killing Docker Swarm. If it can be done before `docker swarm init`, please include it into the upcoming 17.04 release.

**Comment by mavenugo** [comment]
@joelchen no. this can be done only after `docker swarm init`. But once you do that once, it is remembered even after the daemon is restarted.

**Comment by thaJeztah** [comment]
https://github.com/moby/moby/pull/31714 was merged, and released in 17.05, so I'll go ahead and close this issue.
